torch_xla/experimental/fori_loop.py

```diff
@@ -10,80 +10,139 @@
-    return lower[0] < upper[0]
+  def cond_fn(iteri, *input_value):
+    return iteri > 0
 
-  def body_fn(upper, lower, *init_val):
-    one_value_i = torch.ones(1, dtype=torch.int32, device=device)
-    res_list = list(user_body_func(*init_val))
-    res_list.insert(0, lower)
-    res_list.insert(0, torch.sub(upper, one_value_i))
-    return res_list
+  def new_body_fn(iteri, *input_value):
+    return iteri - 1, body_fun(*input_value)
+
+  inputs = (iteri,) + input_value
+  res = _xla_while_loop_wrapper(
+      cond_fn, new_body_fn, inputs, (), fake_tensor=True)
 
-  res = while_loop(cond_fn, body_fn, (lower, upper, *init_val))
   return res
 
 
 @while_loop_op.py_impl(DispatchKey.XLA)
-def while_loop(cond_fn, body_fn, *carried_inputs, additional_inputs=None):
-  # TODO(@manfei): PyTorch require carried_inputs to be list/tuple, PyTorch/XLA _xla_while_loop only accept *operands, *operands would tuple items again: (a, '')
-  # cond_fn&body_fn: callable
-  # carried_inputs: (Tuple of possibly nested dict/list/tuple of tensors)
+def while_loop(cond_fn, body_fn, carried_inputs, additional_inputs=None):
   if additional_inputs is None:
     additional_inputs = tuple()
-  return _xla_while_loop(
-      cond_fn, body_fn, *carried_inputs, additional_inputs=additional_inputs)
+  return _xla_while_loop_wrapper(cond_fn, body_fn, carried_inputs,
+                                 additional_inputs)
+
+
+def _xla_while_loop_wrapper(cond_fn,
+                            body_fn,
+                            carried_inputs,
+                            additional_inputs=None,
+                            fake_tensor=False):
+
+  def new_body_fn(*carried_inputs):
+    res = list(body_fn(*carried_inputs))
+    if additional_inputs:
+      res = [
+          res[0],
+      ] + list(additional_inputs) + res[1:]
+    else:
+      res = res
+    return res
 
+  return _xla_while_loop(cond_fn, new_body_fn, carried_inputs,
+                         additional_inputs, fake_tensor)
 
-def _xla_while_loop(cond_fn, body_fn, *carried_inputs, additional_inputs):
-  # untuple carried_inputs from while_loop
-  carried_inputs = carried_inputs[0]
-  # fake carried_inputs to split formal code
+
+def _xla_while_loop(cond_fn,
+                    body_fn,
+                    carried_inputs,
+                    additional_inputs=None,
+                    fake_tensor=False):
+
+  #  ====== fake_carried_inputs ======
   fake_carried_inputs = []
   for carried_input in carried_inputs:
     device = carried_input.device
     fake_carried_inputs.append(
         torch.randint(10, carried_input.size(),
                       dtype=carried_input.dtype).to(device))
-  fake_carried_inputs = tuple(fake_carried_inputs)
 
-  # trans fake_carried_inputs from list(tensor) to list(xla::op)
-  kwargs = {}
-  if type(fake_carried_inputs) is tuple:
-    shapes = xb.tensor_shape(fake_carried_inputs)
+  fake_input_output = fake_carried_inputs[1:]
+  fake_iter_input = fake_carried_inputs[:-1]
```

Comment:
can you add comments and examples for these two? For example what oder of `carried_input ` do you expect?

Reply:
thanks, added comments/example for expected inputs;

deleted these three lines due to not used

Reply:
thanks for review, rebase missed, updated now